routers/web/repo/editor.go

```diff
@@ -392,6 +392,14 @@ func EditFilePost(ctx *context.Context) {
 // NewFilePost response for creating file
 func NewFilePost(ctx *context.Context) {
 	form := web.GetForm(ctx).(*forms.EditRepoFileForm)
+
+	// trim leading and trailing spaces from the tree path
+	treePath := strings.Split(form.TreePath, "/")
+	for i, v := range treePath {
+		treePath[i] = strings.TrimSpace(v)
+	}
+	form.TreePath = strings.Join(treePath, "/")
```

Comment:
Hmm, the path could be something like `/ / / /`, just press Space Slash Space Slash on the editor page 🤣 

Then the tree path becomes "////", which doesn't seem good.

Reply:
Ah, we are already in `NewFilePost`, so likely above remark is invalid.

Reply:
> Then the tree path becomes "////", which doesn't seem good.

I assume the backend would reject that? Seems like right behaviour to me.

Reply:
There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

Reply:
Actually I think we could just warn the users on the UI, instead of doing too much.

The more we do, the more edge cases come.

Reply:
> There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

Same as entering `..` before and I hope we don't follow this reference in backend.

Not sure about warning, for me it seems clear-cut that a user would never want leading/trailing whitespace in the file or folder names and if they really do, they ought to use a git client.

Reply:
> Hmm, the path could be something like `/ / / /`, just press Space Slash Space Slash on the editor page 🤣
> 
> Then the tree path becomes "////", which doesn't seem good.

In this case, there is prevention by the front-end, we can't just type '/' in the path without any other character.

> There are more edge cases: for example, what happens to `foo/ .. /bar`? Will it be rejected or become `foo/../bar` then become `bar`?

In the current latest version, if the path is ' .. ' with leading or trailing spaces, it will cause some incorrect behavior because '..' with leading and trailing spaces is not allowed in the system.

For example, the test case 't1/ ../t2/test-1' will not be created, and the system will just respond with the same string.

Typing '..' might seem weird, but I believe that people who type '..' do it with meaningful intent.

Reply:
Preventing this on the front-end is a good solution, but currently, we can't see spaces in the UI.

This is the current UI display, but every t-n I have added some leading or trailing space, which is not being shown.
![image](https://github.com/go-gitea/gitea/assets/30816317/093a88ce-d051-483d-8fd8-be93ff6a8fab)

To prevent this on the front-end, we have two choices:

1. Change the display to show spaces. This will let people see what they have typed and display a warning.
2. Just show a warning (I think this is not ideal because people can't see the full typed path in the current UI).

Reply:
Tested with GitHub, I think GitHub also uses my suggestion: do not do too much for end users, only warn them.

Reply:
GitHub has another good behavior: if `../` is inputted, then it changes the path to parent, not sure whether it is implemented in Gitea's UI.

Reply:
> Tested with GitHub, I think GitHub also uses my suggestion: do not do too much for end users, only warn them.在 GitHub 上测试过，我认为 GitHub 也采用了我的建议：不要为最终用户做太多事情，只警告他们。

On GitHub, it shows a warning but does not take any action for users. This is my test: 't2  /  t3   /   t4  /test' 
![圖片](https://github.com/go-gitea/gitea/assets/30816317/28091f06-757d-464d-971e-bf46a263f6d9)
![圖片](https://github.com/go-gitea/gitea/assets/30816317/ef4c843c-4cd0-4b38-b598-2ffffb345ba6)

This problem is the same as #31478.

I'm not sure if we need to follow this behavior or not.

Reply:
If GitHub has this warning, then I agree, we should implement it too.

Reply:
Both approaches sound good to me.
Both get an LGTM from me.

Reply:
![firefox_TrY7rDIQ4d](https://github.com/go-gitea/gitea/assets/30816317/671f22cd-81a9-4a82-bd40-17c7dece7b07)

I added some warnings for users similar to GitHub. 
However, I have kept backend processing for spaces. 
On GitHub, it only removes spaces from filenames, not from parent directories. Since Gitea can run on different operating systems, leading or trailing spaces may cause some problems. Therefore, I have retained space processing.